### ISSUE TITLE
ckeygen --showpub does not prompt for or use provided passphrase for encrypted keys

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5890]
|<img alt="ltaylor.volks's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ltaylor.volks reported|
|-|-|
|Trac ID|trac#5890|
|Type|defect|
|Created|2012-08-20 00:46:13Z|

```
$ ckeygen --showpub -filename=/Users/blah/.ssh/id_rsa --pass=passwurd
Traceback (most recent call last):
  File "/Users/blah/Develop/virtualenvs/lodgeprox/bin/ckeygen", line 7, in <module>
    execfile(__file__)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/bin/conch/ckeygen", line 15, in <module>
    run()
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/scripts/ckeygen.py", line 65, in run
    displayPublicKey(options)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/scripts/ckeygen.py", line 141, in displayPublicKey
    key = keys.Key.fromFile(options['filename']).keyObject
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 63, in fromFile
    return Class.fromString(file(filename, 'rb').read(), type, passphrase)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 91, in fromString
    return method(data, passphrase)
  File "/Users/blah/Develop/code/3rdparty/twisted-hg/twisted/conch/ssh/keys.py", line 228, in _fromString_PRIVATE_OPENSSH
    raise EncryptedKeyError('encrypted key with no passphrase')
twisted.conch.ssh.keys.EncryptedKeyError: encrypted key with no passphrase
```

Same error occurs if no passphrase is provided.

2 problems:

 1. `twisted.conch.scripts.ckeygen.displayPublicKey` is not catching an `EncryptedKeyError` thrown when no passphrase is provided.

 2.  If a passphrase **is** provided, it is only used to find the key if a `BadKeyError` is caught. But `BadKeyError` is not thrown in case of missing passphrase on an encrypted key, an `EncryptedKeyError` is. `twisted.conch.scripts.ckeygen.displayPublicKey` never prompts for or provides a passphrase to `twisted.conch.ssh.keys.Key.fromFile().

Attachments:

* [5890-ckeygen](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6c9/6c9a2d98fd534fbea330f55fe2daec081900ec44/1a80ce1d99e0f9b8ce6f87b30b072be1458d1b33) (13512 bytes) - added by ltaylor.volks on 2012-08-20 01:53:24Z - Fix ckeygen --showpub for encrypted keys; add tests; whitespace cleanup
* [5890.bugfix](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6c9/6c9a2d98fd534fbea330f55fe2daec081900ec44/6817d9e072a54af3bfec38739975f97e14e0aaf2.bugfix) (114 bytes) - added by ltaylor.volks on 2012-08-20 05:11:24Z - Added topfile NEWS entry
* [5890-ckeygen.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6c9/6c9a2d98fd534fbea330f55fe2daec081900ec44/139b7b585d93af8922645ec72c4ba8c8bb4b10cc.diff) (13780 bytes) - added by ltaylor.volks on 2012-08-21 00:50:30Z - Fix ckeygen --showpub for encrypted keys; add tests; whitespace cleanup (SVN diff...ignore previous attachment)

<details><summary>Searchable metadata</summary>

```
trac-id__5890 5890
type__defect defect
reporter__ltaylor_volks ltaylor.volks
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__easy easy
time__1345423573000000 1345423573000000
changetime__1346146783000000 1346146783000000
version__None None
owner__therve therve
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @therve|
|-|-|

Looks good! I'm merging it alongside [#5889](https://github.com/twisted/twisted/issues/5889) with some adjustments.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set status to `closed`|
|-|-|

(In [35420]) Fix ckeygen --showpub behavior, passing a type to display in toString and
catching the proper exception for encrypted keys.

Author: ltaylor.volks
Reviewer: therve
Fixes: [#5889](https://github.com/twisted/twisted/issues/5889), [#5890](https://github.com/twisted/twisted/issues/5890)
